File: python/genParenthesis.py
class Solution(object):
    def generate(self, s, left, right, n, combinations):
        if left==n and right==n:
            combinations.append(s)
        else:
            if left<n:
                self.generate(s+'(', left+1, right, n, combinations)
            if right<left:
                self.generate(s+')', left, right+1, n, combinations)
    
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        combinations = []
        self.generate('',0,0,n, combinations)
        return combinations

File: java/genParenthesis.java
class Solution {
    public void generate(String combo, List<String> combos, int left, int right, int n) {
        if(left==n && right==n) {
            combos.add(combo);
        } else {
            if(left<n)
                generate(combo+"(", combos, left+1, right, n);
            if(right<left)
                generate(combo+")", combos, left, right+1, n);
        }
    }
    
    public List<String> generateParenthesis(int n) {
        ArrayList<String> combos = new ArrayList<>();
        generate("", combos, 0, 0, n);
        return combos;
    }
}